src/code/models/simulation.coffee

```diff
@@ -89,10 +97,49 @@ module.exports = class Simulation
 
     @reportFunc(data)
 
+  # Tests that the graph contains no loops consisting of dependent variables.
+  # A graph such as A->B<->C is invalid if B and C connect to each other and
+  # neither are accumulators
+  graphIsValid: ->
+    nodesSeenInSegment = []
+    _.each @nodes, (node) -> node._isValid = null
+
+    # Recursive function. We go throw a node's non-independent ancestors, and if
+    # we find ourselves again, or if any of our ancestors find themselves again,
+    # we have a loop.
+    isInALoop = (node) ->
+      linksIn = node.inLinks()
+      # quick exit if we're not a dependent variable, or if we've already been checked
+      return false if node._isValid or node.isAccumulator or linksIn.length is 0
+      for seen in nodesSeenInSegment
+        if seen is node
+          return true
+
+      nodesSeenInSegment.push node
+
+      for link in linksIn
+        return true if isInALoop link.sourceNode
+
+      return false
+
+    for node in @nodes
+      nodesSeenInSegment = []
```

Comment:
Actually, this is the critical declaration, where as the one on line 104 shouldn't be there.

It's on 104 as well because of my occasional uncomfortableness of coffeescript -- not being able to place an explicit `var` declaration yourself. To my mind, it's clearer to see `var nodesSeenInSegment` at the top of this function, because it makes it clear that it's declared outside the scope of the inner function, whereas that's not immediately apparent when the initial declaration is below the inner function.

But I should remove it, because it's non-functional. Coffeescript will correctly put the declaration at the top of the function anyway, so I shouldn't worry about it.

(In terms of it's function, which is also maybe a little confusing: nodesSeenInSegment is all the nodes we have encountered when looking at one node's ancestors. This gets cleared for each new node we ask for, although these subsequent lookups for the other nodes ancestors will generally be much faster because many of them may already have been marked as good.)